phonon/update.py

```diff
@@ -234,5 +198,183 @@ def merge(self, update):
+        return self._end_session(self.__execute, self.__cache, block)
+
+    def cache(self):
+        """
+        This method caches the update to redis.
+        """
+        self.process().client.set(self.resource_id,
+                                  pickle.dumps(self))
+
+    def __merge(self):
+        """
+        Handles how to extract a resource's data from redis and calls the user
+        defined `merge` method.
+        """
+        pickled = self.process().client.get(self.resource_id)
+        if pickled:
+            cached = pickle.loads(pickled)
+            self.merge(cached)
+
+    def __cache(self, block=None):
+        """
+        Handles deciding whether or not to get the resource from redis. Also
+        implements merging cached records with this one (by implementing your
+        `merge` method). Increments the number of times this record was
+        modified if the cache method executes successfully (does not raise).
+        """
+        if block is None:
+            block = self.block
+
+        with self.ref.lock(block=block):
+            if self.ref.get_times_modified() > 0:
+                self.__merge()
+            self.cache()
+            self.ref.increment_times_modified()
+
+        if self.init_cache:
+            self._clear()
+
+    def __execute(self, block=None):
+        """
+        Handles deciding whether or not to get the resource from redis. Also
+        implements merging cached records with this one (by implementing your
+        `merge` method). Increments the number of times this record was
+        modified if the cache method executes successfully (does not raise).
+        """
+        if block is None:
+            block = self.block
+
+        with self.ref.lock(block=block):
+            if self.ref.get_times_modified() > 0:
+                self.__merge()
+            self.execute()
+
+
+class ConflictFreeUpdate(BaseUpdate):
+    """ 
+    Update class to be used when looking to minimize the locking of resources
+    during caching and execution of updates.  Accordingly, it should only be
+    used when you can ensure that you are only performing atomic operations on
+    both Redis and your database.
+
+    A default caching and merging method is provided which should be sufficient
+    when simply incrementing or decrementing fields. Otherwise, they must be
+    overwritten to suit your needs.
+    """
+    def __init__(self, *args, **kwargs):
+        super(ConflictFreeUpdate, self).__init__(*args, **kwargs)
+        if self.init_cache:
+            self.__cache()
+
+    def end_session(self, block=True):
+        return self._end_session(self.__execute, self.__cache, block)
+
+    def get_cached_data(self):
+        """
+        Gets the cached data for the resource from redis and reconstructs it
+        from a flattened structure into its original form.
+        """
+        cached_data = self.process().client.hgetall(self.resource_id)
```

Comment:
I would suggest implementing a `__get_cached_state` and `__get_cached_doc` method. I think this might be a little prematurely optimized. We should do our best to avoid generic names like `cached_data` and `reconstructed_data`